es811x/src/main/scala/tech/beshu/ror/es/utils/EsqlRequestHelper.scala

```diff
@@ -0,0 +1,275 @@
+  }
+
+  def modifyIndicesOf(request: CompositeIndicesRequest,
+                      requestTables: NonEmptyList[IndexTable],
+                      finalIndices: Set[String]): Either[ModificationError, CompositeIndicesRequest] = {
+    Try {
+      setQuery(request, newQueryFrom(getQuery(request), requestTables, finalIndices))
+    }.toEither.left.map(ModificationError.UnexpectedException.apply)
+  }
+
+  def modifyResponseAccordingToFieldLevelSecurity(response: ActionResponse,
+                                                  fieldLevelSecurity: FieldLevelSecurity): Either[ModificationError, ActionResponse] = {
+    Try(new EsqlQueryResponse(response).modifyByApplyingRestrictions(fieldLevelSecurity.restrictions).underlyingObject)
+      .toEither.left.map(ModificationError.UnexpectedException.apply)
+  }
+
+  sealed trait EsqlRequestClassification
+  object EsqlRequestClassification {
+    final case class IndicesRelated(tables: NonEmptyList[IndexTable]) extends EsqlRequestClassification {
+      lazy val indices: Set[String] = tables.toCovariantSet.flatMap(_.indices.toIterable)
+    }
+    case object NonIndicesRelated extends EsqlRequestClassification
+  }
+
+  sealed trait ClassificationError
+  object ClassificationError {
+    final case class UnexpectedException(ex: Throwable) extends ClassificationError
+    case object ParsingException extends ClassificationError
+  }
+
+  import EsqlRequestClassification._
+
+  def classifyEsqlRequest(request: CompositeIndicesRequest): Either[ClassificationError, EsqlRequestClassification] = {
+    val result = Try {
+      val query = getQuery(request)
+      val params = ReflecUtils.invokeMethodCached(request, request.getClass, "params")
+
+      implicit val classLoader: ClassLoader = request.getClass.getClassLoader
+      Try(new EsqlParser().createStatement(query, params)) match {
+        case Success(statement: IndicesRelatedStatement) => Right(IndicesRelated(statement.indices))
+        case Success(command: OtherCommand) => Right(NonIndicesRelated)
+        case Failure(_) => Left(ClassificationError.ParsingException: ClassificationError)
+      }
+    }
+    result match {
+      case Success(value) => value
+      case Failure(exception) => Left(ClassificationError.UnexpectedException(exception))
+    }
+  }
+
+  private def getQuery(request: CompositeIndicesRequest): String = {
+    ReflecUtils.invokeMethodCached(request, request.getClass, "query").asInstanceOf[String]
+  }
+
+  private def setQuery(request: CompositeIndicesRequest, newQuery: String): CompositeIndicesRequest = {
+    ReflecUtils
+      .getMethodOf(request.getClass, Modifier.PUBLIC, "query", 1)
+      .invoke(request, newQuery)
+    request
+  }
+
+  private def newQueryFrom(oldQuery: String, requestTables: NonEmptyList[IndexTable], finalIndices: Set[String]) = {
+    requestTables.toList.foldLeft(oldQuery) {
+      case (currentQuery, table) =>
+        val (beforeFrom, afterFrom) = currentQuery.splitBy("FROM")
```

Comment:
There is a tip in docs: `ES|QL keywords are case-insensitive`, so we should be aware of both cases.
It seems that SQL is affected too